lib/puppet/parser/functions.rb

```diff
@@ -30,11 +30,43 @@ def self.reset
     end
   end
 
+  class AutoloaderDelegate
+    attr_reader :delegatee
+
+    def initialize
+      @delegatee = Puppet::Util::Autoload.new(self, "puppet/parser/functions")
+    end
+
+    def loadall(env = Puppet.lookup(:current_environment))
+      if Puppet[:strict] != :off
+        Puppet.warn_once('deprecations', 'Puppet::Parser::Functions', _("The method 'Puppet::Parser::Functions.autoloader#loadall' is deprecated in favor of using 'Scope#call_function'."))
+      end
+
+      @delegatee.loadall(env)
+    end
+
+    def load(name, env = Puppet.lookup(:current_environment))
+      if Puppet[:strict] != :off
+        Puppet.warn_once('deprecations', 'Puppet::Parser::Functions', _("The method 'Puppet::Parser::Functions.autoloader#load(\"%{name}\")' is deprecated in favor of using 'Scope#call_function'.") % {name: name})
+      end
+
+      @delegatee.load(name, env)
+    end
+
+    def loaded?(name)
+      if Puppet[:strict] != :off
+        Puppet.warn_once('deprecations', 'Puppet::Parser::Functions', _("The method 'Puppet::Parser::Functions.autoloader#loaded?(\"%{name}\")' is deprecated in favor of using 'Scope#call_function'.") % {name: name})
+      end
+
```

Comment:
The above three deprecations - is it desired that there is only one warning for any calls into one of the three methods `loadall` `load` and `loaded` - the messages differ so the label should be different I think - e.g. 'Puppet::Parser::Functions#loadall', 'Puppet::Parser::Functions#load', etc.  If not done that way, you only see the first of the three warnings.

Would we perhaps even want a warn_once per file that uses those calls? If so, the caller must be included in the label.

Reply:
Ah that wasn't my intention. I can include the `name` of the function that the caller is trying to load/check to see if it's loaded. Including the caller is possible:

```
caller = File.basename(caller_locations(1,1).first.path, '.*')
```

But it's also stack-dependent and a bit too clever. I'll update the PR to include the method called.